internal/metrics/metrics.go

```diff
@@ -5,13 +5,117 @@ import (
 	"net/http"
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
 const (
 	MetricsEndpoint = "0.0.0.0:9090"
 )
 
+var (
+	EventsCounter *prometheus.CounterVec
+
+	ConditionCounter        *prometheus.CounterVec
+	ConditionRunTimeSummary *prometheus.SummaryVec
+
+	ActionCounter        *prometheus.CounterVec
+	ActionRuntimeSummary *prometheus.SummaryVec
+
+	ActionHandlerCounter        *prometheus.CounterVec
+	ActionHandlerRunTimeSummary *prometheus.SummaryVec
+
+	DownloadBytes          *prometheus.CounterVec
+	DownloadRunTimeSummary *prometheus.SummaryVec
+	UploadBytes            *prometheus.CounterVec
+	UploadRunTimeSummary   *prometheus.SummaryVec
+
+	StoreQueryErrorCount *prometheus.CounterVec
+)
+
+func init() {
+	EventsCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flasher_events_received",
+			Help: "A counter metric to measure the total count of events received",
+		},
+		[]string{"valid", "response"}, // valid is true/false, response is ack/nack
+	)
+
+	ConditionCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flasher_conditions_received",
+			Help: "A counter metric to measure the total count of condition requests received, successful and failed",
+		},
+		[]string{"condition", "state"},
+	)
+
+	ConditionRunTimeSummary = promauto.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "flasher_condition_duration_seconds",
+			Help: "A summary metric to measure the total time spent in completing each condition",
+		},
+		[]string{"condition", "state"},
+	)
+
+	ActionCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flasher_actions_counter",
+			Help: "A counter metric to measure the sum of firmware install actions executed",
+		},
+		[]string{"vendor", "component", "state"},
+	)
+
+	ActionRuntimeSummary = promauto.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "flasher_install_action_runtime_seconds",
+			Help: "A summary metric to measure the total time spent in each install action",
+		},
+		[]string{"vendor", "component", "state"},
+	)
+
+	ActionHandlerCounter = promauto.NewCounterVec(
```

Comment:
Do we need `ActionHandlerCounter`? cf. [This discussion of counts and sums of histograms and summaries](https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations)

Reply:
TIL, thanks, looking at the summary metrics made it clearer to me where its duplicated.

```
flasher_actions_counter{component="bios",state="failed",vendor="dell"} 1
flasher_condition_duration_seconds_sum{condition="firmwareInstall",state="failed"} 161.947962561
flasher_condition_duration_seconds_count{condition="firmwareInstall",state="failed"} 2
flasher_conditions_received{condition="firmwareInstall",state="failed"} 2
flasher_events_received{response="ack",valid="true"} 2
flasher_install_action_handler_counter{component="bios",state="failed",transition="checkingInstalledFirmware",vendor="dell"} 1
flasher_install_action_handler_counter{component="bios",state="succeeded",transition="poweringOnDevice",vendor="dell"} 1
flasher_install_action_handler_seconds_sum{component="bios",state="failed",transition="checkingInstalledFirmware",vendor="dell"} 47.192571911
flasher_install_action_handler_seconds_count{component="bios",state="failed",transition="checkingInstalledFirmware",vendor="dell"} 1
flasher_install_action_handler_seconds_sum{component="bios",state="succeeded",transition="poweringOnDevice",vendor="dell"} 3.722863795
flasher_install_action_handler_seconds_count{component="bios",state="succeeded",transition="poweringOnDevice",vendor="dell"} 1
flasher_install_action_runtime_seconds_sum{component="bios",state="failed",vendor="dell"} 50.926325003
flasher_install_action_runtime_seconds_count{component="bios",state="failed",vendor="dell"} 1
flasher_store_query_error_count{queryKind="GetFirmwareSet",storeKind="serverservice"} 1
```